// sucrase/dist/sucrase-babylon/tokenizer/state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class State {
    init() {
        this.potentialArrowAt = -1;
        this.noAnonFunctionType = false;
        this.tokens = [];
        this.scopes = [];
        this.pos = 0;
        this.type = 2560 /* eof */;
        this.start = this.pos;
        this.end = this.pos;
        this.isType = false;
    }
    snapshot() {
        return {
            potentialArrowAt: this.potentialArrowAt,
            noAnonFunctionType: this.noAnonFunctionType,
            tokensLength: this.tokens.length,
            scopesLength: this.scopes.length,
            pos: this.pos,
            type: this.type,
            contextualKeyword: this.contextualKeyword,
            start: this.start,
            end: this.end,
            isType: this.isType,
        };
    }
    restoreFromSnapshot(snapshot) {
        this.potentialArrowAt = snapshot.potentialArrowAt;
        this.noAnonFunctionType = snapshot.noAnonFunctionType;
        this.tokens.length = snapshot.tokensLength;
        this.scopes.length = snapshot.scopesLength;
        this.pos = snapshot.pos;
        this.type = snapshot.type;
        this.contextualKeyword = snapshot.contextualKeyword;
        this.start = snapshot.start;
        this.end = snapshot.end;
        this.isType = snapshot.isType;
    }
}
exports.default = State;
